python/ray/serve/_private/replica.py

```diff
@@ -812,145 +809,213 @@ async def call_reconfigure(self, user_config: Any):
-            assert len(request_args) == 3
-            scope, receive, send = request_args
+        return request_args, asgi_args, receive_task
 
-            if isinstance(self._callable, ASGIAppReplicaWrapper):
-                request_args = (scope, receive, send)
-            else:
-                request_args = (Request(scope, receive, send),)
-        elif request_metadata.is_grpc_request:
-            # Ensure the request args are a single gRPCRequest object.
-            assert len(request_args) == 1 and isinstance(request_args[0], gRPCRequest)
-            request_args = (pickle.loads(request_args[0].grpc_user_request),)
+    def _prepare_args_for_grpc_request(
+        self,
+        request: gRPCRequest,
+        request_metadata: RequestMetadata,
+        user_method_params: Dict[str, inspect.Parameter],
+    ) -> Tuple[Tuple[Any], Dict[str, Any]]:
+        """Prepare arguments for a user method handling a gRPC request.
 
-        runner_method = None
-        try:
-            runner_method = self._get_user_callable_method(request_metadata.call_method)
-            if inspect.isgeneratorfunction(runner_method) or inspect.isasyncgenfunction(
-                runner_method
-            ):
-                raise TypeError(
-                    f"Method '{runner_method.__name__}' is a generator function. "
-                    "You must use `handle.options(stream=True)` to call "
-                    "generators on a deployment."
-                )
+        Returns (request_args, request_kwargs).
+        """
+        request_args = (pickle.loads(request.grpc_user_request),)
+        if GRPC_CONTEXT_ARG_NAME in user_method_params:
+            request_kwargs = {GRPC_CONTEXT_ARG_NAME: request_metadata.grpc_context}
+        else:
+            request_kwargs = {}
 
-            method_to_call = sync_to_async(runner_method)
+        return request_args, request_kwargs
 
-            # Edge case to support empty HTTP handlers: don't pass the Request
-            # argument if the callable has no parameters.
-            params = inspect.signature(runner_method).parameters
-            if request_metadata.is_http_request and len(params) == 0:
-                request_args, request_kwargs = tuple(), {}
-            elif request_metadata.is_grpc_request and GRPC_CONTEXT_ARG_NAME in params:
-                request_kwargs = {GRPC_CONTEXT_ARG_NAME: request_metadata.grpc_context}
-
-            result = await method_to_call(*request_args, **request_kwargs)
-            if inspect.isgenerator(result) or inspect.isasyncgen(result):
+    async def _handle_user_method_result(
+        self,
+        result: Any,
+        user_method_name: str,
+        request_metadata: RequestMetadata,
+        *,
+        generator_result_callback: Optional[Callable],
+        is_asgi_app: bool,
+        asgi_args: Optional[ASGIArgs],
+    ) -> Any:
+        """Postprocess the result of a user method.
```

Comment:
@shrekris-anyscale @zcin this docstring should help explain the bits that were confusing before PTAL

Reply:
Looks much better, thanks!